database/migrations/default/1746472913390_ems_update_triggers_matched_crash_pks/up.sql

```diff
@@ -0,0 +1,117 @@
+            
+        -- Get the count from the array length (handling when array_length is null as 0)
+        SELECT COALESCE(array_length(matched_crash_ids, 1), 0) INTO match_count;
+
+        -- For records that have been manually matched we want to only update the automated match column
+        IF matching_ems.crash_match_status === 'matched_by_manual_qa' THEN
+          IF match_count = 0 THEN
+            UPDATE ems__incidents 
+            SET matched_crash_pks = NULL
+            WHERE id = matching_ems.id;
+          ELSE 
+            -- if the match count is more than one we will update the matched_crash_pks
+            UPDATE ems__incidents 
+            SET matched_crash_pks = matched_crash_ids
+            WHERE id = matching_ems.id;
+          END IF;
+        -- If the record has not been manually matched
+        ELSE
+          IF match_count = 0 THEN
+              UPDATE ems__incidents 
+              SET crash_pk = NULL,
+                  crash_match_status = 'unmatched',
+                  matched_crash_pks = NULL
+              WHERE id = matching_ems.id;
+          ELSIF match_count = 1 THEN
+              -- this EMS record is only matched to one crash - we can assign the crash_pk and matched_crash_pks
+              UPDATE ems__incidents 
+              SET crash_pk = NEW.id,
+                  crash_match_status = 'matched_by_automation',
+                  matched_crash_pks = matched_crash_ids
+              WHERE id = matching_ems.id;
+          ELSE
+              -- multiple matching crashes found - can assign the matched_crash_pks but not crash_pk
+              UPDATE ems__incidents 
+              SET crash_match_status = 'multiple_matches_by_automation',
+                  crash_pk = NULL,
+                  matched_crash_pks = matched_crash_ids
+              WHERE id = matching_ems.id;
+          END IF;
+        END IF;
+    END LOOP;
+
+    --
+    -- nullify the crash_pk for EMS records that no longer match this crash
+    --
+    IF TG_OP = 'UPDATE' THEN
+        UPDATE ems__incidents
+        SET crash_pk = NULL,
+            crash_match_status = 'unmatched',
+            matched_crash_pks = NULL
+        WHERE crash_pk = NEW.id
+          AND (
+              incident_received_datetime  < (NEW.crash_timestamp - time_threshold)
+              OR incident_received_datetime  > (NEW.crash_timestamp + time_threshold)
+              OR geometry IS NULL
+              OR NEW.position IS NULL
+              OR NEW.is_temp_record is TRUE
+              or NEW.is_deleted is TRUE
+              OR (NOT ST_DWithin(geometry::geography, NEW.position::geography, meters_threshold))
+          )
+          AND 
+            -- ignore manual qa matches: these should not be modified
+            crash_match_status != 'matched_by_manual_qa';
+    END IF;
+    RETURN NEW;
+END;
+$function$;
+
+UPDATE ems__incidents set matched_crash_pks = ARRAY[crash_pk] where crash_match_status = 'matched_by_automation';
```

Comment:
backfilling all the records that have one match through automation to populate the `matched_crash_pks` column with the `crash_pk` column value, so that when a user changes the `crash_pk` they will be able to use the reset button to revert to the automated match

Reply:
Additionally, we are going to have a bunch of `matched_by_manual_qa` records that may have originally been `matched_automatically`. If we wanted to enable resetting for these records, we could write a query that mimics the matching conditions of this trigger. I'm not sure if that's going to be worth it, given how unlikely it is that Xavier would want to reset them 🤔

Reply:
if it's unlikely to happen, they could just manually reset in the future by referencing the change log https://github.com/cityofaustin/atd-data-tech/issues/21714